src/dawn/Support/Casting.h

```diff
@@ -316,6 +316,13 @@ inline typename cast_retty<X, Y*>::ret_type dyn_cast_or_null(Y* Val) {
   return (Val && isa<X>(Val)) ? cast<X>(Val) : nullptr;
 }
 
+/// @brief `dyn_cast` for `std::shared_ptr`s
+/// @ingroup core
+template <class T, class U>
+inline std::shared_ptr<T> dyn_pointer_cast(const std::shared_ptr<U>& ptr) noexcept {
```

Comment:
I dont undestand why we do this, it defeats the whole purpose of the shared_ptr by creating another instance on the same raw pointer?

Reply:
This is so that we can apply our custom RTTI system (base on LLVM's RTTI system http://llvm.org/docs/HowToSetUpLLVMStyleRTTI.html) to `shared_ptr` instead of just raw pointers. Note that this does *not* create a new instance as this constructor is commonly know as alias constructor (see constructor (8) in http://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr)